### PR TITLE
Tutorial: A simple monitoring solution

### PR DESCRIPTION
**Pull request by winged**
Introduce a tutorial for Chirp, to show users roughly how it works. The
tutorial introduces the user via a simple monitoring solution that
consist of the following parts that communicate with each other:

* An "agent" program that periodically checks a TCP service for its
  availability, and reports it via chirp to an upstream node
* A "monitor" program that collects and displays the service's status
* A "collector" program that collects data from multiple agents and
  forwards it to another collector, or a monitor.

The monitor checks upon the service list periodically and notes "agent
down" if it didn't hear from the agent for some time (twice the reported
polling interval).

### CODE REVIEW

File: doc/tutorial/agent.c [hunk 1]
```diff
@@ -1,72 +1,225 @@
...
+    ch_message_t*  msg = malloc(sizeof(msg));
+    mon_service_t* svc = malloc(sizeof(svc));
+
+    ch_msg_init(msg);
+    ch_msg_set_address(msg, AF_INET, upstream_host, upstream_port);
+
+    // Note: This is very naive, and any self-respecting code should use
+    // a proper serializer for this.
+    msg->data     = (void*) svc;
+    msg->data_len = sizeof(svc);
```
Comment (ganwell) [inline doc/tutorial/agent.c]: sizeof(*svc);

File: doc/tutorial/agent.c [hunk 2]
```diff
@@ -1,72 +1,225 @@
...
+
 static
 void
-_ch_tst_sent_cb(
-        ch_chirp_t* chirp,
-        ch_message_t* msg,
-        int status,
-        float load
-)
+_poll_connected(uv_connect_t* conn, int status)
```
Comment (ganwell) [inline doc/tutorial/agent.c]: If you want to adapt libchirp naming conventions this thing would be called _poll_connect_cb. The idea is to denote a callback with cb and then the past tense isn't needed anymore since it is clear that the cb fires on a connected event.

File: doc/tutorial/agent.c [hunk 3]
```diff
@@ -1,72 +1,225 @@
...
 {
-    (void)(chirp);
-    (void)(status);
-    (void)(load);
-    ch_chirp_release_recv_handler(msg);
+    _notify_status(status == 0);
+
+    uv_close((uv_handle_t*)conn, NULL);
```
Comment (ganwell) [inline doc/tutorial/agent.c]: Probably should shutdown the handle.

```c
        uv_shutdown( _shutdown_req,  (uv_stream_t*) &conn, _shutdown_cb);
```
Then the close moves to _shutdown_cb. It don't know if its ok, to just close. Maybe we can try it out in this example. You will get a WARNING or FATAL when closing chirp if stuff is still open.

File: doc/tutorial/agent.c [hunk 4]
```diff
@@ -1,72 +1,225 @@
...
+// .. code-block:: cpp
+
+static
+void
+_do_poll(uv_timer_t* timer)
+{
+    uv_tcp_init(ch_chirp_get_loop(_chirp_instance), &poll_socket);
+    struct sockaddr_in dest;
+
+    // TODO: support ipv6 as well
```
Comment (ganwell) [inline doc/tutorial/agent.c]: For the sake of brevity I would leave that out, the decisions on cleanup vs brevity will be difficult enough.

File: doc/tutorial/agent.c [hunk 5]
```diff
@@ -1,72 +1,225 @@
...
+    uv_ip4_addr(service_host, service_port, &dest);
+    uv_tcp_connect(&poll_connect, &poll_socket, (const struct sockaddr*)&dest, _poll_connected);
+
+    uv_timer_init(ch_chirp_get_loop(_chirp_instance), &poll_timeout_timer);
+    poll_timeout_timer.data = _chirp_instance;
+
+    // We wait for two seconds before we assume the service "down".
+    uv_timer_start(&poll_timeout_timer, _poll_timeout, 2000, 0);
```
Comment (ganwell) [inline doc/tutorial/agent.c]: _poll_timeout_cb? See above

File: doc/tutorial/agent.c [hunk 6]
```diff
@@ -1,72 +1,225 @@
...
+static
+void
+_chirp_started(ch_chirp_t* chirp)
+{
+    uv_timer_init(ch_chirp_get_loop(chirp), &poll_timer);
+    poll_timer.data = chirp;
+
+    uv_timer_start(&poll_timer, _do_poll, 0, poll_interval * 1000);
+
+}
```
Comment (ganwell) [inline doc/tutorial/agent.c]: _chirp_start_cb? See above

File: doc/tutorial/agent.c [hunk 7]
```diff
@@ -1,72 +1,225 @@
...
+//
+// .. code-block:: cpp
+
+static
+void
+_new_message(ch_chirp_t* chirp, ch_message_t* msg)
+{
+    ch_chirp_release_recv_handler(msg);
+}
+
```
Comment (ganwell) [inline doc/tutorial/agent.c]: _new_message_cb?

File: doc/tutorial/agent.c [hunk 8]
```diff
@@ -1,72 +1,225 @@
...
+#include <stdlib.h>
+#include <string.h>
+
+#include "common.h"
 #include "libchirp.h"
 
-// System includes
-// ===============
+
+// Setup a few global data structures
```
Comment (ganwell) [inline doc/tutorial/agent.c]: If you want to be able to reorder functions for better narrative, you need to declare the them first.

File: doc/tutorial/agent.c [hunk 9]
```diff
@@ -1,72 +1,225 @@
...
+
+// Received callback
+// -----------------
+//
+// In the agent, we don't really do anything here.
+//
+// Maybe later we'll respond to health checks from peers
+//
+// .. code-block:: cpp
+
```
Comment (ganwell) [inline doc/tutorial/agent.c]: You can set the callback to NULL for brevity if you want. As long as no messages arrive to error is shown.

File: doc/tutorial/agent.c [hunk 10]
```diff
@@ -79,21 +232,71 @@ main(int argc, char *argv[])
...
+    // TODO: refactor host/port parsing into its own function for shorter code
+    char* upstream_port_str = strstr(upstream_host, ":");
+    if (upstream_port_str == NULL) {
+        fprintf(stderr, "Upstream format must be host:port\n");
         exit(1);
     }
+    upstream_port_str++; // port start AFTER the colon..
+    upstream_port = strtol(upstream_port_str, NULL, 10);
+
+    // Hack - replace colon with NUL byte, so upstream_host becomes
```
Comment (ganwell) [inline doc/tutorial/agent.c]: Wouldn't call that a hack. Seems totally legit.

File: doc/tutorial/collector.c [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+static int   upstream_port;
+
+
+// Startup callback
+// ----------------
+//
+// Chirp needs a function to let us know when it has finished initializing.
+// 
+// TODO: Can we leave this out if it's not doing anything? ie. pass it as NULL
+// to ``ch_chirp_run``?
```
Comment (ganwell) [inline doc/tutorial/collector.c]: You can set any of the callbacks to NULL. The idea with the interface is to make people remember that they should set these callbacks, but you don't have to.

```c
//    :param ch_recv_cb_t recv_cb: Called when chirp receives a message.
//    :param ch_start_cb_t start_cb: Called when chirp is started, can be NULL.
//    :param ch_done_cb_t done_cb: Called when chirp is finished, can be NULL.
//    :param ch_log_cb_t log_cb: Callback to the logging facility, can be NULL.
```

I intentionally left out that the recv callback can also be NULL, since this is 99% of the time a mistake, but actually even it can be NULL.

File: doc/tutorial/collector.c [hunk 12]
```diff
@@ -0,0 +1,170 @@
...
+//
+// .. code-block:: cpp
+static
+void
+_new_message(ch_chirp_t* chirp, ch_message_t* msg)
+{
+    // Set upstream address, send out message
+    ch_msg_set_address(msg, AF_INET, upstream_host, upstream_port);
+    ch_chirp_send(chirp, msg, _sent_cb);
+}
```
Comment (ganwell) [inline doc/tutorial/collector.c]: _new_message_cb?

File: doc/tutorial/collector.c [hunk 13]
```diff
@@ -0,0 +1,170 @@
...
+    if (upstream_port_str == NULL) {
+        fprintf(stderr, "Upstream format must be host:port\n");
+        exit(1);
+    }
+    upstream_port_str++; // port start AFTER the colon..
+    upstream_port = strtol(upstream_port_str, NULL, 10);
+
+    // Hack - replace colon with NUL byte, so upstream_host becomes
+    // a string that only contains the hostname. This is "safe", as
+    // we've checked for the existance of the colon already
```
Comment (ganwell) [inline doc/tutorial/collector.c]: Still no hack :smile:

File: doc/tutorial/monitor.c [hunk 14]
```diff
@@ -0,0 +1,325 @@
...
+// "database". To do this, we must be able to find the corresponding entry, if
+// it exists.
+//
+// We return NULL if it doesn't, so the calling code can create an entry instead.
+//
+// .. code-block:: cpp
+
+static
+mon_service_t*
+_service_get(mon_service_t* msg_svc)
```
Comment (ganwell) [inline doc/tutorial/monitor.c]: Nice, its sad that chirp hide the rbtree header in its internals, otherwise you could have used it.

File: mk/base.mk [hunk 15]
```diff
@@ -109,6 +110,21 @@ doc:
...
 endif
 
+$(BUILD)/doc/tutorial/agent.c.rst: $(BASE)/doc/tutorial/agent.c
+	$(V_M)$(BASE)/mk/c2rst $< $@
+$(BUILD)/doc/tutorial/monitor.c.rst: $(BASE)/doc/tutorial/monitor.c
+	$(V_M)$(BASE)/mk/c2rst $< $@
+$(BUILD)/doc/tutorial/collector.c.rst: $(BASE)/doc/tutorial/collector.c
+	$(V_M)$(BASE)/mk/c2rst $< $@
+$(BUILD)/doc/tutorial/common.h.rst: $(BASE)/doc/tutorial/common.h
+	$(V_M)$(BASE)/mk/c2rst $< $@
```
Comment (ganwell) [inline mk/base.mk]: This should not be needed. Just copy i.e.  $(BASE)/doc/tutorial/common.h to  $(BUILD)/doc/tutorial/common.h

**Comment by ganwell** [comment]
@winged I don't know if everything is reviewed perfectly. Lets do another iteration later.

File: doc/tutorial/agent.c (hunk 3 above)
Comment (ganwell) [reply doc/tutorial/agent.c]: @winged at least after rebasing upstream/master

File: README.rst [hunk 16]
```diff
@@ -104,7 +104,14 @@ Install to prefix /usr, but copy to package dir. (Package creation)
...
+.. _source_dist:
+
 How to create a source distribution
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Building a source distribution is useful when you need to include libchirp in
+your project, but don't want to use it's build system. A source distribution can
+easily be compiled with just a ``make`` call.
 
 .. code-block:: bash
```
Comment (ganwell) [inline README.rst]: @winged I also wrote about the source distribution. Your text is better. But you have t rebase again, sorry!

File: doc/building.rst [hunk 17]
```diff
@@ -0,0 +1,86 @@
+=============================
+Building and installing Chirp
+=============================
+
+Unix
+====
+
+Build dependencies:
+
+* python3 [3]_
```
Comment (ganwell) [inline doc/building.rst]: @winged I changed this a bit in the README.rst because of the source distribution. After rebasing you can probably copy paste that part from README.rst. Sorry again.

File: doc/tutorial/agent.c [hunk 18]
```diff
@@ -1,72 +1,393 @@
...
+    notify_status(status == 0, 1);
+
+    /* Also stop the timer - No need for waiting to see if the connection
+    * times out anymore. */
+    uv_timer_stop(&poll_timeout_timer);
+
+
+    /* We don't really want to talk to the service, so immediately request a
+     * shutdown of the socket..  */
+    uv_shutdown_t* poll_closer = malloc(sizeof(*poll_closer));
```
Comment (ganwell) [inline doc/tutorial/agent.c]: You could have this statically allocated, since there is no overlap. (Only one poll socket) It's not important.

File: doc/tutorial/agent.c [hunk 19]
```diff
@@ -1,72 +1,393 @@
...
+// very busy. Let's report it as "down" in that case (maybe we'll add a
+// timeout status later...)
+//
+// .. code-block:: cpp
+
+static
+void
+connection_timeout_cb(uv_timer_t* timer)
+{
+    notify_status(0, 1);
```
Comment (ganwell) [inline doc/tutorial/agent.c]: You should cleanup the connection here. Or am I missing something?

File: doc/tutorial/agent.c [hunk 20]
```diff
@@ -1,72 +1,393 @@
...
+// This helper is called to notify our upstream about our service's status.
+//
+// .. code-block:: cpp
+
+static void
+notify_status(int svc_status, int agent_status)
+{
+    ch_message_t*  msg = malloc(sizeof(*msg));
+    mon_service_t* svc = malloc(sizeof(*svc));
+
```
Comment (ganwell) [inline doc/tutorial/agent.c]: There is no concurrency here, this could also be statically allocated. Again not important.

File: doc/tutorial/agent.c [hunk 21]
```diff
@@ -1,72 +1,393 @@
...
+// .. code-block:: cpp
+
+static
+void
+new_message_cb(ch_chirp_t* chirp, ch_message_t* msg)
 {
-    (void)(chirp);
-    (void)(status);
-    (void)(load);
     ch_chirp_release_message(msg);
```
Comment (ganwell) [inline doc/tutorial/agent.c]: If you build this with the make file you should get: chirp variable not used. Seems the makefile is broken

File: doc/tutorial/collector.c [hunk 22]
```diff
@@ -0,0 +1,188 @@
...
+// This is needed to clean up after a message has been sent upstream.
+// We use it to release the recv handler from the message that we received.
+//
+// .. code-block:: cpp
+
+static
+void
+sent_cb(ch_chirp_t* chirp, ch_message_t* msg, int status)
+{
+    (void)(status);
```
Comment (ganwell) [inline doc/tutorial/collector.c]: Why no: chirp unused variable?

File: doc/tutorial/common.h [hunk 23]
```diff
@@ -0,0 +1,93 @@
...
+    uint16_t   polling_interval;
+    time_t     last_check;
+    uint8_t    status;
+} mon_service_t ;
+
+// The service object represents a service with it's status, both on the wire
+// and in memory of the monitor service. Next to the service's name, we tell
+// the upstream the polling interval that's being used, and a timestamp of
+// the last time a check has been done. The last part is the service's status as
+// described above.
```
Comment (ganwell) [inline doc/tutorial/common.h]: Maybe tell people that we directly send the struct over the net for brevity and that its generally not a good idea.

File: doc/tutorial/index.rst [hunk 24]
```diff
@@ -0,0 +1,181 @@
...
+.. code-block:: bash
+
+   gcc agent.c libchirp.c -o agent -lssl -luv -lm -lpthread -lcrypto
+   gcc monitor.c libchirp.c -o monitor -lssl -luv -lm -lpthread -lcrypto
+   gcc collector.c libchirp.c -o collector -lssl -luv -lm -lpthread -lcrypto
+
+But with the Makefile in place, just run ``make``, and everything will be
+built. Of course, at this point in time, this won't work just yet (unless you
+cheated and downloaded the source files already!).
+
```
Comment (ganwell) [inline doc/tutorial/index.rst]: Maybe we should build this in the travis run. We make dist in travis anyways, so we can copy the files from the dist and build the tutorial. BUT lets do this in another PR. Otherwise you have to rebase all the time.

File: doc/tutorial/index.rst [hunk 25]
```diff
@@ -0,0 +1,181 @@
...
+To help you get started, we've also created a :download:`Makefile <Makefile>`.
+for you that you can use as a starting base. If you prefer to roll your own,
+you can compile the code with the following GCC flags:
+
+.. code-block:: bash
+
+   gcc agent.c libchirp.c -o agent -lssl -luv -lm -lpthread -lcrypto
+   gcc monitor.c libchirp.c -o monitor -lssl -luv -lm -lpthread -lcrypto
+   gcc collector.c libchirp.c -o collector -lssl -luv -lm -lpthread -lcrypto
```
Comment (ganwell) [inline doc/tutorial/index.rst]: Nice!

File: doc/tutorial/index.rst [hunk 26]
```diff
@@ -0,0 +1,181 @@
...
+peer of an agent "upstream". A collector's upstream can in turn be a collector,
+or the "final" monitor.
+
+For simplicity, we limit the service names to 32 chars, and put some more
+restrictions out, which would need to be removed for a full-blown monitoring
+solution. But for the scope of this tutorial, it's more than enough.
+
+Monitor
+-------
+
```
Comment (ganwell) [inline doc/tutorial/index.rst]: ?

File: doc/tutorial/monitor.c [hunk 27]
```diff
@@ -0,0 +1,529 @@
...
+{
+    // Try to find the service in our list of services
+
+    if (_services == NULL) {
+        return NULL;
+    }
+    for (size_t i = 0; i < _services_size; i++) {
+        if (strncmp(_services[i].name, msg_svc->name, 32) == 0) {
+            // Found! Return the service
+
```
Comment (ganwell) [inline doc/tutorial/monitor.c]: Empty lines? I first didn't want to say anything. But then I realized you sometime do them, sometimes not. I personally would prefer less empty lines. But that's a matter of taste.

File: doc/tutorial/monitor.c [hunk 28]
```diff
@@ -0,0 +1,529 @@
...
+{
+
+    mon_service_t* internal_svc = service_get(msg_svc);
+    if (internal_svc == NULL) {
+        // Nothing to do, service does not exist anymore
+        return;
+    }
+
+    /* Special care needs to be taken: If removing the last element,
+     * we don't want to copy the last element over itself. */
```
Comment (ganwell) [inline doc/tutorial/monitor.c]: Instances where you don't do the empty lines. Remove them in the other places?

File: doc/tutorial/monitor.c [hunk 29]
```diff
@@ -0,0 +1,529 @@
...
+// .. code-block:: cpp
+
+static
+void
+new_message_cb(ch_chirp_t* chirp, ch_message_t* msg)
+{
+    /* Note: This makes some very naive assumptions regarding the peer.  Any
+     * respectable network code should NEVER EVER do this! But it keeps the
+     * tutorial short, as we don't need code to serialize/unserialize. */
+    mon_service_t* svc = (mon_service_t*) msg->data;
```
Comment (ganwell) [inline doc/tutorial/monitor.c]: Ok here is the comment about not sending struct over the net. Sorry about the message above.

File: doc/tutorial/monitor.c [hunk 30]
```diff
@@ -0,0 +1,529 @@
...
+show_service_status(mon_service_t *svc, int show_header)
+{
+    char* format = "%-32s %22s %10s %-15s\n";
+
+    char interval_buf[30];
+    char lastcheck_buf[30];
+    struct tm lastcheck;
+
+    if (show_header) {
+        printf(format, "Service", "Last update", "Interval", "State");
```
Comment (ganwell) [inline doc/tutorial/monitor.c]: Just that you know, you don't get errors from the compiler about printf if you use a format variable. Maybe paste it once to check if everything is ok and then revert back to the format variable.

File: doc/tutorial/monitor.c [hunk 31]
```diff
@@ -0,0 +1,529 @@
...
+void
+sig_handler_cb(uv_signal_t* handle, int signum)
+{
+    /* First, we stop the sanity check timer. It's not needed anymore */
+    uv_timer_stop(&poll_timer);
+
+    /* Then, free the services list */
+    free(_services);
+    _services_size = 0;
+    ch_chirp_close_ts(_chirp_instance);
```
Comment (ganwell) [inline doc/tutorial/monitor.c]: This is not complete cleanup. I propose either just die() or cleanup everything.

File: doc/tutorial/monitor.c [hunk 32]
```diff
@@ -0,0 +1,529 @@
...
+// and start it, telling it which callback to run when a message arrives..
+//
+// .. code-block:: cpp
+
+int
+main(int argc, char *argv[])
+{
+    /* Parse the commandline arguments. The monitor is the simplest of all,
+     * just accepting a single port number to listen on.
+     */
```
Comment (ganwell) [inline doc/tutorial/monitor.c]: A flag. Close comment on the same line (not important)

**Comment by winged** [comment]
Okay, merging as discussed via phone. The remaining review items will be fixed in a follow up PR.
